src/views/lighttable.c

```diff
@@ -1770,9 +1772,59 @@ void scrolled(dt_view_t *self, double x, double y, int up, int state)
   }
 }
 
+void activate_control_element(dt_view_t *self)
+{
+  dt_library_t *lib = (dt_library_t *)self->data;
+  switch(lib->image_over)
+  {
+    case DT_VIEW_DESERT:
+    {
+      int32_t id = dt_control_get_mouse_over_id();
+      if((lib->modifiers & (GDK_SHIFT_MASK | GDK_CONTROL_MASK)) == 0)
+        dt_selection_select_single(darktable.selection, id);
+      else if((lib->modifiers & (GDK_CONTROL_MASK)) == GDK_CONTROL_MASK)
+        dt_selection_toggle(darktable.selection, id);
+      else if((lib->modifiers & (GDK_SHIFT_MASK)) == GDK_SHIFT_MASK)
+        dt_selection_select_range(darktable.selection, id);
+
+      break;
+    }
+    case DT_VIEW_REJECT:
+    case DT_VIEW_STAR_1:
+    case DT_VIEW_STAR_2:
+    case DT_VIEW_STAR_3:
+    case DT_VIEW_STAR_4:
+    case DT_VIEW_STAR_5:
+    {
+      int32_t mouse_over_id = dt_control_get_mouse_over_id();
+      dt_image_t *image = dt_image_cache_get(darktable.image_cache, mouse_over_id, 'w');
+      if(image)
+      {
+        if(lib->image_over == DT_VIEW_STAR_1 && ((image->flags & 0x7) == 1))
+          image->flags &= ~0x7;
+        else if(lib->image_over == DT_VIEW_REJECT && ((image->flags & 0x7) == 6))
+          image->flags &= ~0x7;
+        else
+        {
+          image->flags &= ~0x7;
+          image->flags |= lib->image_over;
+        }
+        dt_image_cache_write_release(darktable.image_cache, image, DT_IMAGE_CACHE_SAFE);
+      }
+      else
+        dt_image_cache_write_release(darktable.image_cache, image, DT_IMAGE_CACHE_RELAXED);
+      _update_collected_images(self);
+      break;
+    }
+    default:
+      break;
+  }
+}
 
 void mouse_moved(dt_view_t *self, double x, double y, double pressure, int which)
 {
+  dt_library_t *lib = (dt_library_t *)self->data;
+  lib->activate_on_release = 0;
```

Comment:
Would it make sense to record the x/y position in `button_pressed` and compare that to the final one in `button_released` instead of handling mouse movement here? That way one could easily allow tiny movements which might be impossible to avoid with some touch devices, pens or touchpads.

Reply:
What you say makes sense to me. I see two ways to proceed:

1 - We do not pan the view until the mouse pointer exits the UI element which is activated. If the button is released before leaving, then it counts as a click. As soon as the pointer leaves we start to pan.

2 - Quick and dirty: if the pointer moves less than, say, 5px it's a click.

I would go for the first option, but if you prefer the second I can also implement that one. What's your opinion?

Reply:
The first sounds good, but I would have to see it in action to get a feel for how annoying it is when the view doesn't move at first. Maybe the 2nd will turn out to be the better approach after all. Instead of using a fixed distance we should pass the (say) 5px through `DT_PIXEL_APPLY_DPI()` to make it a real world distance on the screen.